### ISSUE TITLE
Gateway 4.0.4.224: Text received as BLOB when connecting to an Echo service through a Proxy

### ISSUE THREAD

**Issue by sbadugu** [kaazing/gateway#154]
<p><b>Client technology:</b> Javascript<br/>
<b>Browser with Version:</b> tested with Chrome and FF<br/>
<b>Client O/S with Version:</b> Ubuntu 12.04<br/>
<b>Same-origin / Cross-origin?</b> n/a</p>

<p><b>Gateway with version:</b> Gateway 4.0.4.224<br/>
<b>Gateway Java version:</b> 1.7<br/>
<b>Server OS with version:</b> Ubuntu 12.04<br/>
<b>Gateway Service:</b> Proxy/Echo</p>

<p><b>Back-End System with version:</b> n/a</p>

<p><b>Can the bug be reproduced in a Kaazing demo out-of-the-box?</b> No, use attached configuration files.</p>

<p><b>Does the bug reproduce with all emulated or native combinations?</b></p>

<p><b>Is this a regression from previous released version?</b></p>

<p><b>Does this reproduce in a later released version?</b></p>

<p><b>Gateway config, if applicable:</b> attached</p>

<p><b>Client code to reproduce, if applicable:</b> OOB JS demo client</p>

<p><b>Steps to reproduce:</b></p>

<ol>
    <li>Start Gateways using attached configuration file</li>
    <li>Browse to <a href="http://localhost:8001/demo/gateway/javascript/ws.html" class="external-link" rel="nofollow">http://localhost:8001/demo/gateway/javascript/ws.html</a></li>
    <li>Connect to ws://localhost:8004/proxy and click "Send Text"</li>
</ol>

<p><b>Observed behavior:</b></p>

<ol>
    <li>Demo console shows
<p/>
<div id="syntaxplugin" class="syntaxplugin" style="border: 1px dashed #bbb; border-radius: 5px !important; overflow: auto; max-height: 30em;">
<table cellspacing="0" cellpadding="0" border="0" width="100%" style="font-size: 1em; line-height: 1.4em !important; font-weight: normal; font-style: normal; color: black;">
        <tbody  >
                <tr id="syntaxplugin_code_and_gutter">
                        <td  style=" line-height: 1.4em !important; padding: 0em; vertical-align: top;">
                    <pre style="font-size: 1em; margin: 0 10px;  margin-top: 10px;   width: auto; padding: 0;"><span style="color: black; font-family: 'Consolas', 'Bitstream Vera Sans Mono', 'Courier New', Courier, monospace !important;">CONNECT: ws://localhost:8004/outer</span></pre>
            </td>
        </tr>
                <tr id="syntaxplugin_code_and_gutter">
                        <td  style=" line-height: 1.4em !important; padding: 0em; vertical-align: top;">
                    <pre style="font-size: 1em; margin: 0 10px;   width: auto; padding: 0;"><span style="color: black; font-family: 'Consolas', 'Bitstream Vera Sans Mono', 'Courier New', Courier, monospace !important;">CONNECTED</span></pre>
            </td>
        </tr>
                <tr id="syntaxplugin_code_and_gutter">
                        <td  style=" line-height: 1.4em !important; padding: 0em; vertical-align: top;">
                    <pre style="font-size: 1em; margin: 0 10px;   width: auto; padding: 0;"><span style="color: black; font-family: 'Consolas', 'Bitstream Vera Sans Mono', 'Courier New', Courier, monospace !important;">SEND TEXT: Hello, WebSocket!</span></pre>
            </td>
        </tr>
                <tr id="syntaxplugin_code_and_gutter">
                        <td  style=" line-height: 1.4em !important; padding: 0em; vertical-align: top;">
                    <pre style="font-size: 1em; margin: 0 10px;   margin-bottom: 10px;  width: auto; padding: 0;"><span style="color: black; font-family: 'Consolas', 'Bitstream Vera Sans Mono', 'Courier New', Courier, monospace !important;">RECEIVED BLOB: 72,101,108,108,111,44,32,87,101,98,83,111,99,107,101,116,33
</span></pre>
            </td>
        </tr>
            </tbody>
</table>
</div>
<p/></li>
</ol>

<p><b>Expected behavior:</b></p>

<p/>

<div id="syntaxplugin" class="syntaxplugin" style="border: 1px dashed #bbb; border-radius: 5px !important; overflow: auto; max-height: 30em;">
<table cellspacing="0" cellpadding="0" border="0" width="100%" style="font-size: 1em; line-height: 1.4em !important; font-weight: normal; font-style: normal; color: black;">
        <tbody  >
                <tr id="syntaxplugin_code_and_gutter">
                        <td  style=" line-height: 1.4em !important; padding: 0em; vertical-align: top;">
                    <pre style="font-size: 1em; margin: 0 10px;  margin-top: 10px;   width: auto; padding: 0;"><span style="color: black; font-family: 'Consolas', 'Bitstream Vera Sans Mono', 'Courier New', Courier, monospace !important;">CONNECT: ws://localhost:8001/echo</span></pre>
            </td>
        </tr>
                <tr id="syntaxplugin_code_and_gutter">
                        <td  style=" line-height: 1.4em !important; padding: 0em; vertical-align: top;">
                    <pre style="font-size: 1em; margin: 0 10px;   width: auto; padding: 0;"><span style="color: black; font-family: 'Consolas', 'Bitstream Vera Sans Mono', 'Courier New', Courier, monospace !important;">CONNECTED</span></pre>
            </td>
        </tr>
                <tr id="syntaxplugin_code_and_gutter">
                        <td  style=" line-height: 1.4em !important; padding: 0em; vertical-align: top;">
                    <pre style="font-size: 1em; margin: 0 10px;   width: auto; padding: 0;"><span style="color: black; font-family: 'Consolas', 'Bitstream Vera Sans Mono', 'Courier New', Courier, monospace !important;">SEND TEXT: Hello, WebSocket!</span></pre>
            </td>
        </tr>
                <tr id="syntaxplugin_code_and_gutter">
                        <td  style=" line-height: 1.4em !important; padding: 0em; vertical-align: top;">
                    <pre style="font-size: 1em; margin: 0 10px;   margin-bottom: 10px;  width: auto; padding: 0;"><span style="color: black; font-family: 'Consolas', 'Bitstream Vera Sans Mono', 'Courier New', Courier, monospace !important;">RECEIVED TEXT: Hello, WebSocket!
</span></pre>
            </td>
        </tr>
            </tbody>
</table>
</div>

<p/>

<p>Additional notes:<br/>
If you click on "Send Array Buffer" then on "Send Text" you'll see the console say:</p>

<p/>

<div id="syntaxplugin" class="syntaxplugin" style="border: 1px dashed #bbb; border-radius: 5px !important; overflow: auto; max-height: 30em;">
<table cellspacing="0" cellpadding="0" border="0" width="100%" style="font-size: 1em; line-height: 1.4em !important; font-weight: normal; font-style: normal; color: black;">
        <tbody  >
                <tr id="syntaxplugin_code_and_gutter">
                        <td  style=" line-height: 1.4em !important; padding: 0em; vertical-align: top;">
                    <pre style="font-size: 1em; margin: 0 10px;  margin-top: 10px;   width: auto; padding: 0;"><span style="color: black; font-family: 'Consolas', 'Bitstream Vera Sans Mono', 'Courier New', Courier, monospace !important;">SEND TEXT: Hello, WebSocket!</span></pre>
            </td>
        </tr>
                <tr id="syntaxplugin_code_and_gutter">
                        <td  style=" line-height: 1.4em !important; padding: 0em; vertical-align: top;">
                    <pre style="font-size: 1em; margin: 0 10px;   margin-bottom: 10px;  width: auto; padding: 0;"><span style="color: black; font-family: 'Consolas', 'Bitstream Vera Sans Mono', 'Courier New', Courier, monospace !important;">RECEIVED ARRAY BUFFER: 72,101,108,108,111,44,32,87,101,98,83,111,99,107,101,116,33
</span></pre>
            </td>
        </tr>
            </tbody>
</table>
</div>

<p/>

<p>And the same for Byte Buffer.</p>

---

Imported from JIRA
Originally reported by: mustafarian
Issue Number: KG-10527

**Comment by sbadugu** [comment]
**jitendra.kotamraju** said:
<p>I think this is same as <a href="http://jira.kaazing.wan/browse/KG-10275" class="external-link" rel="nofollow">http://jira.kaazing.wan/browse/KG-10275</a></p>

**Comment by sbadugu** [comment]
**chris.barrow** said:
<p>Agreed, I just closed <a href="http://jira.kaazing.wan/browse/KG-10275" title="WS proxy downstream sends text frame as binary frame" class="issue-link" data-issue-key="KG-10275"><del>KG-10275</del></a> as duplicated by this one. Best way to tackle this bug would be to start by creating a robot test (in gateway.server) that reproduces the problem.</p>

<p>Note that WsBuffer does have a flag to indicate text or binary, and we use this in the echo service to make sure we don't lose the type information. Probably when we transit a message through the proxy service we are losing that information (I think that code is in AbstractProxyHandler).</p>

**Comment by sbadugu** [comment]
**brennan.gaunce** said:
<p>Please make sure that the test is for the <b>service</b> only, and does not include transports.</p>

<p>The transport tests are already defined in the "base-service" branch - with a ticket assigned to Jitu.  I'd like to see all of this land in trunk in 4.0.5.  See <a href="http://jira.kaazing.wan/browse/KG-10274" title="Merge base service transport integration tests" class="issue-link" data-issue-key="KG-10274">KG-10274</a>.</p>

**Comment by sbadugu** [comment]
**jitendra.kotamraju** said:
<p>Due to <a href="http://jira.kaazing.wan/browse/KG-10594" title="Unnecessary encoding code " class="issue-link" data-issue-key="KG-10594">KG-10594</a>, WsFrameTextFilter gets added which converts Text frames to binary frames. </p>

**Comment by sbadugu** [comment]
**jitendra.kotamraju** said:
<p>We need more discussion on how a proxy should behave. client --&gt; ws --&gt; tcp, does the tcp layer receive entire websocket frame or just the payload of the websocket frames ? Similarly, what should happen in other cases like, client --&gt; ws --&gt; ws ? Based on the discussion, we need to see how the text and binary frames need to preserved.</p>

**Comment by sbadugu** [comment]
This will need a lot of design change, so we will not fix it in 5.0 + 4.x release.

**Comment by sbadugu** [comment]
gateway-config.xml

```

<gateway-config xmlns="http://xmlns.kaazing.com/2012/08/gateway">
    
    <service>
      <name>extras-directory</name>
      <description>Directory service for demo (extras) files</description>
      <accept>http://localhost:8001/</accept>

      <type>directory</type>
      <properties>
          <directory>/extras</directory>
          <welcome-file>index.html</welcome-file>
          <error-pages-directory>/error-pages</error-pages-directory>
          <options>indexes</options>
      </properties>

  </service>
  <service>
      <accept>ws://localhost:8004/proxy</accept>
      <connect>ws://localhost:8006</connect>
      <type>proxy</type>
      <cross-site-constraint>
          <allow-origin>*</allow-origin>
      </cross-site-constraint>
  </service>
  
  <service>
      <accept>ws://localhost:8006/</accept>
      <type>echo</type>
      <cross-site-constraint>
          <allow-origin>*</allow-origin>
      </cross-site-constraint>
  </service>
</gateway-config>
```
